bee-trace/bee-trace/src/subscriber/mod.rs

```diff
@@ -0,0 +1,146 @@
+    /// into [`tracing`] events, and therefore any external functionality that deals with [`log`] records
+    /// may no longer function as expected.
+    ///  - This method does *not* set the global subscriber. As such, a call to `finish` can be used to
+    /// further extend the return subscriber with external [`Layer`](tracing_subscriber::Layer)s.
+    pub fn finish(mut self) -> Result<(BeeSubscriber, Option<Flamegrapher>), Error> {
+        let (flamegraph_layer, flamegrapher) = self.build_flamegraph_layer()?;
+        let log_layer = self.build_log_layer()?;
+
+        let subscriber = tracing_subscriber::registry().with(flamegraph_layer).with(log_layer);
+
+        Ok((subscriber, flamegrapher))
+    }
+
+    /// Builds the [`BeeSubscriber`] and sets it as the global default subscriber.
+    ///
+    /// Returns a `Result` over an [`Option<Flamegrapher>`](Flamegrapher). The returned option is `Some` if
+    /// the [`LogLayer`](layer::LogLayer) is enabled and has been successfully initialised. If the
+    /// [`LogLayer](layer::LogLayer) has not been enabled with the builder, it is fine to ignore this value.
+    ///
+    /// # Errors
+    ///  - Creation of the [`FlamegraphFilteredLayer`](layer::FlamegraphFilteredLayer) failed.
+    ///  - Creation of the [`LogLayer`](layer::LogLayer) failed.
+    ///
+    /// # Panics
+    /// This method will panic if the flamegraph layer is enabled and the program is not built with
+    /// `--cfg tokio_unstable`.
+    ///
+    /// # Notes
+    ///  - This method calls the [`collect_logs`] function. Any [`log`] records emitted will be converting
+    /// into [`tracing`] events, and therefore any external functionality that deals with [`log`] records
+    /// may no longer function as expected.
+    ///  - This method sets the global subscriber. Any further attempts to set the global subscriber
+    /// (including another call to this method) will fail.
+    ///  - The subscriber initialised by this method cannot be extended.
+    pub fn init(mut self) -> Result<Option<Flamegrapher>, Error> {
+        let (flamegraph_layer, flamegrapher) = self.build_flamegraph_layer()?;
+        let log_layer = self.build_log_layer()?;
+
+        let subscriber = tracing_subscriber::registry().with(flamegraph_layer).with(log_layer);
+
+        subscriber.init();
+
+        Ok(flamegrapher)
+    }
+
+    fn build_log_layer(&mut self) -> Result<Option<layer::LogLayer>, Error> {
+        if self.logger_config.is_some() {
+            collect_logs().map_err(|err| Error::LogLayer(err.into()))?;
+        }
+
+        self.logger_config
+            .take()
+            .map(layer::log_layer)
+            .map_or(Ok(None), |res| res.map(Some))
+    }
+
+    fn build_flamegraph_layer(
+        &mut self,
+    ) -> Result<(Option<layer::FlamegraphFilteredLayer>, Option<Flamegrapher>), Error> {
+        self.flamegraph_stack_file
+            .take()
+            .map(|stack_filename| layer::flamegraph_layer(stack_filename))
```

Comment:
Would it be too bad to put the contents of this map inside the `map_or` instead?